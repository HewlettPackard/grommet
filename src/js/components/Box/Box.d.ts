import * as React from "react";
import Grommet from '../index';

export declare namespace BoxTypes {
  type Size = 'small' | 'medium' | 'large';
  type NoneSize = 'none';
  type Alignment = 'start' | 'center' | 'end' | 'baseline' | 'stretch';
  type AlignContent = 'start' | 'center' | 'end' | 'between' | 'around' | 'stretch';
  type AlignSelf = 'start' | 'center' | 'end' | 'stretch';
  type AnimationType = 'fadeIn' | 'fadeOut' | 'jiggle' | 'pulse' | 'slideUp' | 'slideDown' | 'slideLeft' | 'slideRight' | 'zoomIn' | 'zoomOut';
  interface AnimationObject {
    type?: AnimationType;
    delay?: number;
    duration?: number;
    size?: 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
  }
  type Basis = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge' | 'full' | '1/2' | '1/3' | '2/3' | '1/4' | '3/4';
  type BorderType = 'top' | 'left' | 'bottom' | 'right' | 'horizontal' | 'vertical' | 'all';
  interface BorderObject {
    color: string;
    side: BorderType,
    size: Size
  }
  type Border = BorderType | BorderObject;
  type Direction = 'row' | 'column';
  type Elevation = 'none' | 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
  type Flex = 'grow' | 'shrink' | boolean;
  type Fill = 'horizontal' | 'vertical' | boolean;
  type Gap = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
  type Justify = 'start' | 'center' | 'between' | 'end';
  type JustifySelf = 'start' | 'center' | 'end' | 'stretch';
  type Overflow = 'auto' | 'hidden' | 'scroll';
  type RoundSize = 'xsmall' | 'small' | 'medium' | 'large' | 'full';
  interface BackgroundObject {
    color: string;
    dark: boolean;
    image: string;
    position: string;
    opacity: 'weak' | 'medium' | 'strong' | boolean;
  }
  interface SizeObject {
    bottom: Size;
    horizontal: Size;
    left: Size;
    right: Size;
    top: Size;
    vertical: Size;
  }
}

export interface BoxProps extends Grommet.Props {
  a11yTitle?: string;
  align?: BoxTypes.Alignment;
  alignContent?: BoxTypes.AlignContent;
  alignSelf?: BoxTypes.AlignSelf;
  animation?: BoxTypes.AnimationObject;
  background?: string | BoxTypes.BackgroundObject;
  basis?: BoxTypes.Basis;
  border?: BoxTypes.Border;
  direction?: BoxTypes.Direction;
  elevation?: BoxTypes.Elevation;
  flex?: BoxTypes.Flex;
  fill?: BoxTypes.Fill;
  gap?: BoxTypes.Gap;
  gridArea?: string;
  justify?: BoxTypes.Justify;
  justifySelf?: BoxTypes.JustifySelf;
  margin?: BoxTypes.Size | BoxTypes.SizeObject | BoxTypes.NoneSize;
  overflow?: BoxTypes.Overflow;
  pad?: BoxTypes.Size | BoxTypes.SizeObject | BoxTypes.NoneSize;
  responsive?: boolean;
  round?: BoxTypes.RoundSize;
  tag?: string;
  wrap?: boolean;
}

export class Box extends React.Component<BoxProps, undefined> { }

export default Box;
