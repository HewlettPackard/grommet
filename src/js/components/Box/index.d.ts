import * as React from "react";

export interface BoxProps {
  a11yTitle?: string;
  align?: "start" | "center" | "end" | "baseline" | "stretch";
  alignContent?: "start" | "center" | "end" | "between" | "around" | "stretch";
  alignSelf?: "start" | "center" | "end" | "stretch";
  animation?: "fadeIn" | "fadeOut" | "jiggle" | "pulse" | "slideUp" | "slideDown" | "slideLeft" | "slideRight" | "zoomIn" | "zoomOut" | {type: "fadeIn" | "fadeOut" | "jiggle" | "pulse" | "slideUp" | "slideDown" | "slideLeft" | "slideRight" | "zoomIn" | "zoomOut",delay: number,duration: number,size: "xsmall" | "small" | "medium" | "large" | "xlarge"} | "fadeIn" | "fadeOut" | "jiggle" | "pulse" | "slideUp" | "slideDown" | "slideLeft" | "slideRight" | "zoomIn" | "zoomOut" | {type: "fadeIn" | "fadeOut" | "jiggle" | "pulse" | "slideUp" | "slideDown" | "slideLeft" | "slideRight" | "zoomIn" | "zoomOut",delay: number,duration: number,size: "xsmall" | "small" | "medium" | "large" | "xlarge"}[];
  background?: string | {color: string,dark: boolean | string,image: string,position: string,opacity: "weak" | "medium" | "strong" | boolean,light: string};
  basis?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | "full" | "1/2" | "1/3" | "2/3" | "1/4" | "3/4" | "auto";
  border?: boolean | "top" | "left" | "bottom" | "right" | "horizontal" | "vertical" | "all" | {color: string | {dark: string,light: string},side: "top" | "left" | "bottom" | "right" | "horizontal" | "vertical" | "all",size: "xsmall" | "small" | "medium" | "large"};
  direction?: "row" | "column" | "row-responsive";
  elevation?: "none" | "xsmall" | "small" | "medium" | "large" | "xlarge";
  flex?: "grow" | "shrink" | "true" | "false";
  fill?: "horizontal" | "vertical" | "true" | "false";
  gap?: "xsmall" | "small" | "medium" | "large" | "xlarge";
  gridArea?: string;
  height?: "xsmall" | "small" | "medium" | "large" | "xlarge";
  justify?: "start" | "center" | "between" | "end";
  margin?: "none" | "xsmall" | "small" | "medium" | "large" | "xlarge" | {bottom: "xsmall" | "small" | "medium" | "large" | "xlarge",horizontal: "xsmall" | "small" | "medium" | "large" | "xlarge",left: "xsmall" | "small" | "medium" | "large" | "xlarge",right: "xsmall" | "small" | "medium" | "large" | "xlarge",top: "xsmall" | "small" | "medium" | "large" | "xlarge",vertical: "xsmall" | "small" | "medium" | "large" | "xlarge"};
  overflow?: "auto" | "hidden" | "scroll" | "visible";
  pad?: "none" | "xsmall" | "small" | "medium" | "large" | "xlarge" | {bottom: "xsmall" | "small" | "medium" | "large" | "xlarge",horizontal: "xsmall" | "small" | "medium" | "large" | "xlarge",left: "xsmall" | "small" | "medium" | "large" | "xlarge",right: "xsmall" | "small" | "medium" | "large" | "xlarge",top: "xsmall" | "small" | "medium" | "large" | "xlarge",vertical: "xsmall" | "small" | "medium" | "large" | "xlarge"};
  responsive?: boolean;
  round?: "xsmall" | "small" | "medium" | "large" | "full";
  tag?: string;
  width?: "xsmall" | "small" | "medium" | "large" | "xlarge";
  wrap?: boolean;
}

declare const Box: React.ComponentType<BoxProps>;

export { Box };
