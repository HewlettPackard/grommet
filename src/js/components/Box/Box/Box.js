import React from 'react';
import PropTypes from 'prop-types';
import { Box as BoxM } from '../Box';

function Box(props) {
  return <BoxM {...props} />
}

/* PropTypes for UXPin Merge */
Box.propTypes = {
  children: PropTypes.node,
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(["start", "center", "end", "stretch"]),
  gridArea: PropTypes.string,
  margin: PropTypes.oneOf(["none", "xxsmall", "xsmall", "small", "medium", "large", "xlarge"]),
  align: PropTypes.oneOf(["start", "center", "end", "baseline", "stretch"]),
  alignContent: PropTypes.oneOf(["start", "center", "end", "between", "around", "stretch"]),
  animation: PropTypes.oneOf(["fadeIn", "fadeOut", "jiggle", "pulse", "slideUp", "slideDown", "slideLeft", "slideRight", "zoomIn", "zoomOut"]),
  background: PropTypes.string,
  basis: PropTypes.oneOf(["xxsmall", "xsmall", "small", "medium", "large", "xlarge", "xxlarge", "full", "1/2", "1/3", "2/3", "1/4", "2/4", "3/4", "auto"]),
  border: PropTypes.oneOf([ "top", "left", "bottom", "right", "horizontal", "vertical",  "all"]),
  direction: PropTypes.oneOf(["row", "column", "row-responsive"]),
  elevation: PropTypes.oneOf(["none", "xsmall", "small", "medium", "large", "xlarge"]),
  flex: PropTypes.oneOf(["grow", "shrink"]),
  fill: PropTypes.oneOf(["horizontal", "vertical"]),
  gap: PropTypes.oneOf(["xxsmall", "xsmall", "small", "medium", "large", "xlarge"]),
  height: PropTypes.oneOf(["xxsmall", "xsmall", "small", "medium", "large", "xlarge", "xxlarge"]),
  justify: PropTypes.oneOf(["start", "center", "between", "around", "evenly", "end"]),
  overflow: PropTypes.oneOf(["auto", "hidden", "scroll", "visible"]),
  pad: PropTypes.oneOf(["none", "xxsmall", "xsmall", "small", "medium", "large", "xlarge"]),
  responsive: PropTypes.bool,
  round: PropTypes.bool,
  tag: PropTypes.string,
  as: PropTypes.string,
  width: PropTypes.oneOf(["xxsmall", "xsmall", "small", "medium", "large", "xlarge", "xxlarge"]),
  wrap: PropTypes.bool,
}

/* Export Default for UXPin Merge  */
export default Box;