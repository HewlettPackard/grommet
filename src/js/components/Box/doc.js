import { schema, PropTypes } from 'react-desc';

export default Box => schema(Box, {
  description: `A flexible box that lays out its contents along a single
    direction.`,
  usage: `import { Box } from 'grommet';
  <Box/>`,
  props: {
    align: [
      PropTypes.oneOf(['start', 'center', 'end', 'baseline', 'stretch']),
      'How to align the contents along the cross axis.',
    ],
    alignContent: [
      PropTypes.oneOf(['start', 'center', 'end', 'between', 'around', 'stretch']),
      `How to align the contents when there is extra space in the cross axis.
      Defaults to stretch`,
    ],
    alignSelf: [
      PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
      `How to align along the cross axis when contained in a Box or along
      the column axis when contained in a Grid.`,
    ],
    background: [
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          color: PropTypes.string,
          dark: PropTypes.bool,
          image: PropTypes.string,
          opacity: PropTypes.oneOfType([
            PropTypes.oneOf(['weak', 'medium', 'strong']),
            PropTypes.bool,
          ]),
        }),
      ]),
      `Either a color identifier to use for the background color. For example:
      'neutral-1'. Or, a 'url()' for an image. Dark is not needed if color is provided.`,
    ],
    basis: [
      PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge',
        'full', '1/2', '1/3', '2/3', '1/4', '3/4']),
      'A fixed or relative size along its container\'s main axis.',
    ],
    border: [
      PropTypes.oneOfType([
        PropTypes.oneOf(['top', 'left', 'bottom', 'right',
          'horizontal', 'vertical', 'all']),
        PropTypes.shape({
          color: PropTypes.string,
          side: PropTypes.oneOf(['top', 'left', 'bottom', 'right',
            'horizontal', 'vertical', 'all']),
          size: PropTypes.oneOf(['small', 'medium', 'large']),
          radius: PropTypes.oneOf(['small', 'medium', 'large', 'xlarge']),
        }),
      ]),
      'Include a border.',
    ],
    direction: [
      PropTypes.oneOf(['row', 'column']),
      'The orientation to layout the child components in. Defaults to column.',
    ],
    flex: [
      PropTypes.oneOf(['grow', 'shrink', true, false]),
      'Whether flex-grow and/or flex-shrink is true.',
    ],
    full: [
      PropTypes.oneOf(['horizontal', 'vertical', true, false, 'grow']),
      'Whether the width and/or height should take the full viewport size.',
    ],
    gridArea: [
      PropTypes.string,
      'The name of the area to place this Box in inside a parent Grid.',
    ],
    justify: [
      PropTypes.oneOf(['start', 'center', 'between', 'end']),
      'How to align the contents along the main axis.',
    ],
    justifySelf: [
      PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
      'How to align along the row axis when contained in a Grid.',
    ],
    margin: [
      PropTypes.oneOfType([
        PropTypes.oneOf(['none', 'small', 'medium', 'large']),
        PropTypes.shape({
          bottom: PropTypes.oneOf(['small', 'medium', 'large']),
          horizontal: PropTypes.oneOf(['small', 'medium', 'large']),
          left: PropTypes.oneOf(['small', 'medium', 'large']),
          right: PropTypes.oneOf(['small', 'medium', 'large']),
          top: PropTypes.oneOf(['small', 'medium', 'large']),
          vertical: PropTypes.oneOf(['small', 'medium', 'large']),
        }),
      ]),
      `The amount of margin around the box. An object can be specified to
      distinguish horizontal margin, vertical margin, and margin on a
      particular side of the box`,
    ],
    pad: [
      PropTypes.oneOfType([
        PropTypes.oneOf(['none', 'small', 'medium', 'large']),
        PropTypes.shape({
          bottom: PropTypes.oneOf(['small', 'medium', 'large']),
          horizontal: PropTypes.oneOf(['small', 'medium', 'large']),
          left: PropTypes.oneOf(['small', 'medium', 'large']),
          right: PropTypes.oneOf(['small', 'medium', 'large']),
          top: PropTypes.oneOf(['small', 'medium', 'large']),
          vertical: PropTypes.oneOf(['small', 'medium', 'large']),
        }),
      ]),
      `The amount of padding around the box contents. An object can be specified to
      distinguish horizontal padding, vertical padding, and padding on a
      particular side of the box`,
    ],
    reverse: [
      PropTypes.bool,
      'Whether to reverse the order of the child components.',
    ],
    round: [
      PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'full']),
      'How much to round the corners.',
    ],
    tag: [
      PropTypes.string,
      'The DOM tag to use for the element. Defaults to div.',
    ],
    textAlign: [
      PropTypes.oneOf(['start', 'center', 'end']),
      'How to align the text inside the box.',
    ],
    wrap: [
      PropTypes.bool,
      'Whether children can wrap if they can\'t all fit. Defaults to false.',
    ],
  },
});
