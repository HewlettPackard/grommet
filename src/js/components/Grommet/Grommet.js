import React, { forwardRef, useEffect, useMemo, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  getBreakpoint,
  getDeviceBreakpoint,
  normalizeColor,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const deviceResponsive = (userAgent, theme) => {
  // log('--deviceResponsive', userAgent, theme);
  /*
   * Regexes provided for mobile and tablet detection are meant to replace
   * a full-featured specific library due to contributing a considerable size
   * into the bundle.
   *
   * User agents found https://deviceatlas.com/blog/list-of-user-agent-strings
   */
  if (userAgent) {
    if (/(tablet|ipad|playbook|silk)|(android(?!.*mobile))/i.test(userAgent)) {
      return getDeviceBreakpoint('tablet', theme);
    }
    if (/Mobile|iPhone|Android/.test(userAgent)) {
      return getDeviceBreakpoint('phone', theme);
    }
    return getDeviceBreakpoint('computer', theme);
  }
  return undefined;
};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    ...rest
  } = props;

  const { background, dir, themeMode, userAgent } = props;

  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';
    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(window.innerWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <ContainerTargetContext.Provider value={containerTarget}>
          <StyledGrommet full={full} {...rest} ref={ref}>
            {children}
          </StyledGrommet>
          {full && <FullGlobalStyle />}
        </ContainerTargetContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';

let GrommetDoc;
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  GrommetDoc = require('./doc').doc(Grommet);
}
const GrommetWrapper = GrommetDoc || Grommet;

export { GrommetWrapper as Grommet };
