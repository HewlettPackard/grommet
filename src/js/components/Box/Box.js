import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'recompose';

import { colorForName, colorIsDark } from '../../utils';

import { withTheme } from '../hocs';

import StyledBox, { StyledBoxGap } from './StyledBox';

import doc from './doc';

const styledComponents = {
  div: StyledBox,
}; // tag -> styled component

class Box extends Component {
  static contextTypes = {
    grommet: PropTypes.object,
  }
  static childContextTypes = {
    grommet: PropTypes.object,
  }

  static defaultProps = {
    direction: 'column',
    margin: 'none',
    pad: 'none',
    tag: 'div',
  };

  getChildContext() {
    const { grommet } = this.context;
    const { background, theme } = this.props;
    let dark = false;
    if (background) {
      if (typeof background === 'object') {
        dark = background.dark;
      } else {
        const color = colorForName(background, theme);
        if (color) {
          dark = colorIsDark(color);
        }
      }
      return {
        grommet: { ...grommet, dark },
      };
    }
    return {};
  }

  render() {
    const {
      a11yTitle,
      children,
      direction,
      gap,
      tag,
      theme,
      ...rest
    } = this.props;

    let StyledComponent = styledComponents[tag];
    if (!StyledComponent) {
      StyledComponent = StyledBox.withComponent(tag);
      styledComponents[tag] = StyledComponent;
    }

    let contents = children;
    if (gap) {
      contents = [];
      let firstIndex;
      Children.forEach(children, (child, index) => {
        if (child) {
          if (firstIndex === undefined) {
            firstIndex = index;
          } else {
            contents.push((
              <StyledBoxGap
                key={index}
                gap={gap}
                direction={direction}
                theme={theme}
              />
            ));
          }
        }
        contents.push(child);
      });
    }

    return (
      <StyledComponent
        aria-label={a11yTitle}
        direction={direction}
        theme={theme}
        {...rest}
      >
        {contents}
      </StyledComponent>
    );
  }
}

if (process.env.NODE_ENV !== 'production') {
  doc(Box);
}

export default compose(
  withTheme,
)(Box);
