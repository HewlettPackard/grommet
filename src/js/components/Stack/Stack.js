import React, { cloneElement, Component } from 'react';
import { compose } from 'recompose';

import { withTheme } from '../hocs';

import StyledStack from './StyledStack';
import doc from './doc';
import styleMap from './styleMap';

class Stack extends Component {
  render() {
    const { anchor, children, ...rest } = this.props;

    // make all children but the first absolutely positioned
    const styledChildren = React.Children.map(children, (child, index) => {
      if (index === 0) {
        return child;
      }

      if (child) {
        const style = {
          position: 'absolute',
          overflow: 'hidden',
          ...styleMap[anchor || 'fill'],
        };
        return cloneElement(child, { style });
      }

      return child;
    });

    return (
      <StyledStack {...rest}>
        {styledChildren}
      </StyledStack>
    );
  }
}

if (process.env.NODE_ENV !== 'production') {
  doc(Stack);
}

export default compose(
  withTheme,
)(Stack);
