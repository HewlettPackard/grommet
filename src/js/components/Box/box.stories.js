import React, { Component } from 'react';
import { storiesOf } from '@storybook/react';
import { Attraction, Car, TreeOption } from 'grommet-icons';

import Box from '../Box/Box';
import Text from '../Text/Text';
import Anchor from '../Anchor/Anchor';
import Button from '../Button/Button';
import Grommet from '../Grommet/Grommet';

class SimpleBox extends Component {
  render() {
    return (
      <Grommet>
        <Box
          direction='row-responsive'
          justify='center'
          align='center'
          pad='xlarge'
          background='dark-2'
        >
          <Box
            pad='xlarge'
            align='center'
            background={{ color: 'white-2', opacity: 'strong' }}
          >
            <Attraction size='xlarge' />
            <Text>Party</Text>
            <Anchor href='' label='Link' />
            <Button label='Button' onClick={() => {}} />
          </Box>
          <Box
            pad='xlarge'
            align='center'
            background={{ color: 'accent-2', opacity: 'weak' }}
          >
            <TreeOption size='xlarge' />
            <Text>Nature</Text>
            <Anchor href='' label='Link' />
            <Button label='Button' onClick={() => {}} />
          </Box>
          <Box pad='xlarge' align='center' background='dark-3'>
            <Car size='xlarge' color='light-2' />
            <Text>Travel</Text>
            <Anchor href='' label='Link' />
            <Button label='Button' onClick={() => {}} />
          </Box>
        </Box>
      </Grommet>
    );
  }
}

const customColorBox = {
  global: {
    colors: {
      'brand-gradient': 'linear-gradient(102.77deg, #865ED6 -9.18%, #18BAB9 209.09%)',
    },
  },
};

class CustomColorBox extends Component {
  render() {
    return (
      <Grommet theme={customColorBox}>
        <Box
          justify='center'
          align='center'
          pad='xlarge'
          background={{ color: 'brand-gradient', dark: true }}
          round='large'
        >
          <Text>I have a linear gradient background</Text>
        </Box>
      </Grommet>
    );
  }
}

storiesOf('Box', module)
  .add('Simple Box', () => <SimpleBox />)
  .add('Custom color', () => <CustomColorBox />);
