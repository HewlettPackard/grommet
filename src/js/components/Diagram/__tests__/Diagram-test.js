import React from 'react';
import 'jest-styled-components';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Grommet } from '../../Grommet';
import { Stack } from '../../Stack';
import { Box } from '../../Box';
import { Diagram } from '../';

Enzyme.configure({ adapter: new Adapter() });

const Context = ({ children }) => (
  <Grommet>
    <Stack>
      <Box direction='row'>
        <Box id='1' pad='medium' />
        <Box id='2' pad='medium' />
      </Box>
      {children}
    </Stack>
  </Grommet>
);

test('Diagram renders', (done) => {
  const component = mount(
    <Context>
      <Diagram
        connections={[{ fromId: '1', toId: '2' }]}
      />
    </Context>
  );
  // delay a bit so we can render twice
  setTimeout(() => {
    expect(component.getDOMNode()).toMatchSnapshot();
    component.unmount();
    done();
  }, 10);
});

test('Diagram type renders', () => {
  const component = mount(
    <Context>
      <Diagram
        connections={[
          { fromId: '1', toId: '2', type: 'direct' },
          { fromId: '1', toId: '2', type: 'curved' },
          { fromId: '1', toId: '2', type: 'rectilinear' },
        ]}
      />
    </Context>
  );
  expect(component.getDOMNode()).toMatchSnapshot();
});

test('Diagram color renders', () => {
  const component = mount(
    <Context>
      <Diagram
        connections={[
          { fromId: '1', toId: '2', color: 'brand' },
        ]}
      />
    </Context>
  );
  expect(component.getDOMNode()).toMatchSnapshot();
});

test('Diagram offset renders', (done) => {
  const component = mount(
    <Context>
      <Diagram
        connections={[
          { fromId: '1', toId: '2', offset: 'xsmall' },
          { fromId: '1', toId: '2', offset: 'small' },
          { fromId: '1', toId: '2', offset: 'medium' },
        ]}
      />
    </Context>
  );
  // delay a bit so we can render twice
  setTimeout(() => {
    expect(component.getDOMNode()).toMatchSnapshot();
    component.unmount();
    done();
  }, 10);
});

test('Diagram thickness renders', () => {
  const component = mount(
    <Context>
      <Diagram
        connections={[
          { fromId: '1', toId: '2', thickness: 'xsmall' },
          { fromId: '1', toId: '2', thickness: 'small' },
          { fromId: '1', toId: '2', thickness: 'medium' },
        ]}
      />
    </Context>
  );
  expect(component.getDOMNode()).toMatchSnapshot();
});
