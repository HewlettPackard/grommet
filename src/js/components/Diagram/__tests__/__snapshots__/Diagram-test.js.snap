// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagram color renders 1`] = `
<div
  class="StyledGrommet-gGRXwz deqpVW"
>
  <div
    class="StyledStack-jqCQsC ewABiA"
  >
    <div
      class="StyledBox-bStriS kkBTIr"
      direction="row"
      style="position: relative;"
    >
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="1"
      />
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="2"
      />
    </div>
    <svg
      class="StyledDiagram-fAdbMP bwxUiT"
      viewBox="0 0 0 0"
      preserveAspectRatio="xMinYMin meet"
      width="0"
      height="0"
      style="position: absolute; overflow: hidden; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <g />
    </svg>
  </div>
</div>
`;

exports[`Diagram offset renders 1`] = `
<div
  class="StyledGrommet-gGRXwz deqpVW"
>
  <div
    class="StyledStack-jqCQsC ewABiA"
  >
    <div
      class="StyledBox-bStriS kkBTIr"
      direction="row"
      style="position: relative;"
    >
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="1"
      />
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="2"
      />
    </div>
    <svg
      class="StyledDiagram-fAdbMP bwxUiT"
      viewBox="0 0 0 0"
      preserveAspectRatio="xMinYMin meet"
      width="0"
      height="0"
      style="position: absolute; overflow: hidden; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <g />
    </svg>
  </div>
</div>
`;

exports[`Diagram renders 1`] = `
<div
  class="StyledGrommet-gGRXwz deqpVW"
>
  <div
    class="StyledStack-jqCQsC ewABiA"
  >
    <div
      class="StyledBox-bStriS kkBTIr"
      direction="row"
      style="position: relative;"
    >
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="1"
      />
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="2"
      />
    </div>
    <svg
      class="StyledDiagram-fAdbMP bwxUiT"
      viewBox="0 0 0 0"
      preserveAspectRatio="xMinYMin meet"
      width="0"
      height="0"
      style="position: absolute; overflow: hidden; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <g />
    </svg>
  </div>
</div>
`;

exports[`Diagram thickness renders 1`] = `
<div
  class="StyledGrommet-gGRXwz deqpVW"
>
  <div
    class="StyledStack-jqCQsC ewABiA"
  >
    <div
      class="StyledBox-bStriS kkBTIr"
      direction="row"
      style="position: relative;"
    >
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="1"
      />
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="2"
      />
    </div>
    <svg
      class="StyledDiagram-fAdbMP bwxUiT"
      viewBox="0 0 0 0"
      preserveAspectRatio="xMinYMin meet"
      width="0"
      height="0"
      style="position: absolute; overflow: hidden; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <g />
    </svg>
  </div>
</div>
`;

exports[`Diagram type renders 1`] = `
<div
  class="StyledGrommet-gGRXwz deqpVW"
>
  <div
    class="StyledStack-jqCQsC ewABiA"
  >
    <div
      class="StyledBox-bStriS kkBTIr"
      direction="row"
      style="position: relative;"
    >
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="1"
      />
      <div
        class="StyledBox-bStriS jGLyaD"
        direction="column"
        id="2"
      />
    </div>
    <svg
      class="StyledDiagram-fAdbMP bwxUiT"
      viewBox="0 0 0 0"
      preserveAspectRatio="xMinYMin meet"
      width="0"
      height="0"
      style="position: absolute; overflow: hidden; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <g />
    </svg>
  </div>
</div>
`;
